scripts/configure

```diff
@@ -0,0 +1,254 @@
+}
+
+function gn_generate() { # [project options]
+    mkdir -p "${BUILD_ROOT}"
+    ensure_no_clobber "${BUILD_ROOT}/args.gn"
+    (
+        cd "${CHIP_ROOT}/${PROJECT}" # --root= doesn't work for gn args!
+
+        # Run gn gen with an empty args.gn first so we can list all arguments
+        info "Configuring gn build arguments"
+        echo "# ${CONFIGURE_MARKER}" >"${BUILD_ROOT}/args.gn"
+        gn -q gen "$BUILD_ROOT"
+
+        # Use the argument list to drive the mapping of our command line options to GN args
+        python_util process_project_args <(gn args "$BUILD_ROOT" --list --json) "$@" >>"${BUILD_ROOT}/args.gn"
+
+        # Now gn gen with the arguments we have configured.
+        info "Running gn gen to generate ninja files"
+        gn -q gen "$BUILD_ROOT"
+    )
+}
+
+function create_ninja_wrapper() {
+    # Note: "." != $BUILD_DIR for in-tree builds
+    ensure_no_clobber "ninja"
+    cat >"ninja" <<END
+#!/bin/bash -e
+# ${CONFIGURE_MARKER}
+cd "\$(dirname "\$0")"
+source "${BUILD_ENV_DIR}/bin/activate"
+exec "${NINJA}" -C "${BUILD_DIR}" "\$@"
+END
+    chmod a+x "ninja"
+}
+
+function activate_build_env() {
+    generate_build_env_cksum # re-used by finalize_build_env
+    [[ -r "${BUILD_ENV_DIR}/.cksum" ]] || return 1
+    read -r <"${BUILD_ENV_DIR}/.cksum" || true
+    [[ "$REPLY" == "$CURRENT_ENV_CKSUM" ]] || return 1
+
+    [[ -r "${BUILD_ENV_DIR}/bin/activate" ]] || return 1
+    info "Using existing build environment: ${BUILD_ENV_DIR}"
+    source "${BUILD_ENV_DIR}/bin/activate"
+    PYTHON="python"
+}
+
+function configure_python_env() {
+    progress "Setting up Python venv"
+    "$PYTHON" -m venv "$BUILD_ENV_DIR"
+    info "ok"
+
+    progress "Installing Python build dependencies"
+    "${BUILD_ENV_DIR}/bin/pip" install --disable-pip-version-check --require-virtualenv --use-pep517 --quiet \
+        -r "${CHIP_ROOT}/scripts/setup/requirements.build.txt" \
+        -c "${CHIP_ROOT}/scripts/setup/constraints.txt"
+    info "ok"
+}
+
+function generate_build_env_cksum() {
+    # Conservatively assume that any change to this script or BUILD_ENV_DEPS invalidates the environment
+    CURRENT_ENV_CKSUM="$(cat "$0" "${BUILD_ENV_DEPS[@]}" | cksum)"
+    [[ -n "$CURRENT_ENV_CKSUM" ]] || fail "Failed to generate build environment checksum"
+}
+
+function finalize_build_env() {
+    echo "$CURRENT_ENV_CKSUM" >"${BUILD_ENV_DIR}/.cksum"
+    source "${BUILD_ENV_DIR}/bin/activate"
+    PYTHON="python"
+}
+
+function download_zap() {
+    local version platform
+    read -r version <"${CHIP_ROOT}/scripts/setup/zap.version"
+    case "$OSTYPE" in
+    linux*) platform=linux ;;
+    darwin*) platform=mac ;;
+    *) fail "Unable to determine zap platform for OSTYPE='$OSTYPE'" ;;
+    esac
+    local url="https://github.com/project-chip/zap/releases/download/${version}/zap-${platform}.zip"
+
+    progress "Installing zap-cli from $url"
+    python_util download_and_extract_zip "$url" "${BUILD_ENV_DIR}/bin" zap-cli
```

Comment:
since you go to python anyway, you can perform all the more complex logic there, including having parsing of a json version instead of trying to have a separate file because it is easier for shell to read.

Reply:
It's the stripping of the leading `2@` and trailing `.1` from the CIPD tag that made me go for a separate version file, those CIPD specifics don't really need to pollute everything else. The extra file doesn't really cost us anything since we already have zap_update.py anyway.